design/ingressroute-design.md

```diff
@@ -0,0 +1,401 @@
+
+Each _root_ of a DAG starts at a virtual host, which describes properties like the fully qualified name of the virtual host, any aliases (for example a www. prefix) of that vhost, TLS configuration, and possibly global access list details.
+The vertices of a graph do not contain virtual host information, they are only reachable from a root via a delegation.
+This permits the _owner_ of an ingress root to both delegate the authority to publish a service on a portion of the route space inside a virtual host, and to itself further delegate authority to publish and delegate.
+
+This also means that 
+
+In practice the linkage, or delegation, from root to vertex, is performed with a specific type of route action.
+You can think of it that rather than routing traffic to a service, it is routed to another ingressroute object for further processing.
+
+## Status
+
+Given the large set of validation and delegation errors that are possible the current status of the object should be reported in a top level status object.
+
+TODO(dfc) design a top level `status:` key to capture error information for the user.
+
+# Detailed Design
+
+## IngressRoute CRD
+
+This is an example of a fully populated root ingressroute.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google
+  namespace: prod
+spec:
+  # virtualhost appears at most once. If it is present, the object is considered
+  # to be a "root".
+  virtualhost:
+    hostnames:
+      # each of the fully qualified domain names by which this root should be known as
+      # are listed here. 
+      - www.google.com
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # hostnames, the tls.secretName secret must contain a matching certificate.
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # services defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    services:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
```

Comment:
Would delegating to a non-existent namespace result in the IngressRoute considered invalid? Seems like that state is somewhat of a limbo, which makes me wonder what the status of the IngressRoute would be set to.

Reply:
> Would delegating to a non-existent namespace result in the IngressRoute be considered invalid?

Yes, and No. Delegating a DNS name to a nameserver that is offline or not configured to answer for that domain is not going to get you a functioning DNS record, but it isn't a configuration failure of the upstream nameserver.